File: 프로그래머스/lv3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<vector<int>> triangle) {
    int answer = 0;
    vector<vector<int>> sum(triangle.size(), vector<int>(triangle.size(), 0));
    sum[0][0] = triangle[0][0];
    
    for(int i = 0; i < triangle.size() - 1; i++)
        for(int j = 0; j < triangle[i].size(); j++){
            sum[i+1][j] = max(sum[i+1][j], sum[i][j] + triangle[i+1][j]);
            sum[i+1][j+1] = max(sum[i+1][j+1], sum[i][j] + triangle[i+1][j+1]);
        }
    
    return answer = *max_element(sum[sum.size()-1].begin(), sum[sum.size()-1].end());
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.Arrays;

class Solution {
    public int solution(int[][] triangle) {
        int answer = 0;
        int n = triangle.length;
        int[][] dp = new int[n][n];
        dp[0][0] = triangle[0][0];
        
        for(int i = 0; i < n - 1; i++){
            for(int j = 0; j < triangle[i].length; j++){
                dp[i + 1][j] = Math.max(dp[i][j] + triangle[i + 1][j], dp[i + 1][j]);
                dp[i + 1][j + 1] = Math.max(dp[i][j] + triangle[i + 1][j + 1], dp[i + 1][j +  1]);
            }
        }
        
        return answer = Arrays.stream(dp[n - 1]).max().getAsInt();
    }
}